### PR TITLE
Implement automatic SSH Key Pair creation

### PR DESCRIPTION
**Pull request by SweetOps**
## what
* Automatically generate SSH key pair from `tls_private_key`

## why
* Simplify master key generation

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  algorithm = "${var.ssh_key_algorithm}"
+}
+
+resource "aws_key_pair" "empty" {
+  count      = "${signum(length(var.ssh_public_key_path)) == 0 ? 1 : 0}"
+  key_name   = "${module.label.id}"
+  public_key = "${tls_private_key.default.public_key_openssh}"
+}
+
+resource "null_resource" "save_ssh_keys" {
```
Comment (osterman) [inline main.tf]: Let's rename to `generate_ssh_keys`

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  tags       = "${var.tags}"
+}
+
+resource "aws_key_pair" "non_empty" {
+  count      = "${signum(length(var.ssh_public_key_path)) == 1 ? 1 : 0}"
+  key_name   = "${module.label.id}"
+  public_key = "${file("${var.ssh_public_key_path}")}"
+}
+
+resource "tls_private_key" "default" {
```
Comment (osterman) [inline main.tf]: nice! I did not know about this resource type.

Comment (SweetOps) [reply main.tf]: I did not know it too before today

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  key_name   = "${module.label.id}"
+  public_key = "${tls_private_key.default.public_key_openssh}"
+}
+
+resource "null_resource" "save_ssh_keys" {
+  count      = "${signum(length(var.ssh_public_key_path)) == 0 ? 1 : 0}"
+  depends_on = ["tls_private_key.default"]
+
+  provisioner "local-exec" {
+    command = "echo \"${tls_private_key.default.public_key_openssh}\" > ${path.module}/ssh_key.pub && chmod 600 ${path.module}/ssh_key.pub"
```
Comment (osterman) [inline main.tf]: Output path should be configurable by the caller.

Comment (osterman) [inline main.tf]: Output filename should be `aws_key_pair`'s `key_name`

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,33 @@
+variable "namespace" {
+  default = ""
+}
+
+variable "stage" {
+  default = ""
```
Comment (osterman) [inline variables.tf]: Remove defaults

File: variables.tf [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+variable "name" {
+  default = ""
```
Comment (osterman) [inline variables.tf]: Remove defaults

File: variables.tf [hunk 6]
```diff
@@ -0,0 +1,33 @@
+variable "namespace" {
+  default = ""
```
Comment (osterman) [inline variables.tf]: Remove default (as this is required)

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  provisioner "local-exec" {
+    command = "echo \"${tls_private_key.default.private_key_pem}\" > ${path.module}/ssh_key"
```
Comment (osterman) [inline main.tf]: `path.module` should be replaced with `var.ssh_public_key_path`

**Comment by osterman** [comment]
- [ ] Update `LICENSE`
- [ ] Add `README.md`

File: main.tf (hunk 7 above)
Comment (SweetOps) [reply main.tf]: in this case we'll get the error "file not found" in previous step

File: README.md [hunk 8]
```diff
@@ -11,14 +11,16 @@ module "default" {
...
   name                = "${var.name}"
   ssh_public_key_path = "${var.ssh_public_key_path}"
+  generate_ssh_key    = "${var.generate_ssh_key}"
 }
 ```
 
 ## Variables
 
 |  Name                        |  Default       |  Description                                            | Required |
 |:-----------------------------|:--------------:|:--------------------------------------------------------|:--------:|
```
Comment (osterman) [inline README.md]: Fix whitespace alignment. It looks off.

File: README.md [hunk 9]
```diff
@@ -2,23 +2,25 @@
 
 Terraform module for importing SSH public key file into AWS.
 
-# Usage
+## Usage
 
 ```terraform
 module "default" {
-  source = "git::https://github.com/cloudposse/tf_key_pair.git?ref=tags/0.1.0"
+  source              = "git::https://github.com/cloudposse/tf_key_pair.git?ref=tags/0.1.0"
```
Comment (comeanother) [inline README.md]: Replace `tags/0.1.0` with `master`

File: README.md [hunk 10]
```diff
@@ -2,23 +2,25 @@
...
 |:-----------------------------|:--------------:|:--------------------------------------------------------|:--------:|
-| namespace                    | ``             | Namespace (e.g. `cp` or `cloudposse`)                   | Yes      |
-| stage                        | ``             | Stage (e.g. `prod`, `dev`, `staging`)                   | Yes      |
-| name                         | ``             | Name  (e.g. `bastion` or `db`)                          | Yes      |
-| ssh_public_key_path          | ``             | Path to Read/Write SSH Public Key File (directory)      | Yes      |
+| `namespace`                  | ``             | Namespace (e.g. `cp` or `cloudposse`)                   | Yes      |
+| `stage`                      | ``             | Stage (e.g. `prod`, `dev`, `staging`)                   | Yes      |
+| `name`                       | ``             | Name  (e.g. `bastion` or `db`)                          | Yes      |
+| `ssh_public_key_path`        | ``             | Path to Read/Write SSH Public Key File (directory)      | Yes      |
+| `generate_ssh_key`           | `false`        |  Set to `true` if need generate ssh-keys                | No       |
```
Comment (comeanother) [inline README.md]: Update description.
If set to `true`, new ssh key pair will be created.

File: README.md [hunk 11]
```diff
@@ -2,23 +2,25 @@
...
 |:-----------------------------|:--------------:|:--------------------------------------------------------|:--------:|
-| namespace                    | ``             | Namespace (e.g. `cp` or `cloudposse`)                   | Yes      |
-| stage                        | ``             | Stage (e.g. `prod`, `dev`, `staging`)                   | Yes      |
-| name                         | ``             | Name  (e.g. `bastion` or `db`)                          | Yes      |
-| ssh_public_key_path          | ``             | Path to Read/Write SSH Public Key File (directory)      | Yes      |
+| `namespace`                  | ``             | Namespace (e.g. `cp` or `cloudposse`)                   | Yes      |
+| `stage`                      | ``             | Stage (e.g. `prod`, `dev`, `staging`)                   | Yes      |
+| `name`                       | ``             | Name  (e.g. `bastion` or `db`)                          | Yes      |
+| `ssh_public_key_path`        | ``             | Path to Read/Write SSH Public Key File (directory)      | Yes      |
+| `generate_ssh_key`           | `false`        | If set to `true`, new ssh key pair will be created      | No       |
```
Comment (comeanother) [inline README.md]: Disambiguate
https://github.com/cloudposse/tf_key_pair/pull/1/files#diff-c9ac8098c5ea9d3e6a9a596ff0c512a4R26

**Comment by SweetOps** [comment]
screenshot 2017-09-05 13 27 07

File: README.md (hunk 11 above)
Comment (SweetOps) [reply README.md]: has been fixed

File: output.tf [hunk 12]
```diff
@@ -1,3 +1,7 @@
 output "key_name" {
-  value = "${aws_key_pair.default.key_name}"
+  value = "${aws_key_pair.non_empty.key_name}"
```
Comment (osterman) [inline output.tf]: rewrite like this `${element(compact(aws_key_pair_non_empty.*.key_name, aws_key_pair_empty.*.key_name), 0)}`

File: output.tf [hunk 13]
```diff
@@ -1,3 +1,7 @@
...
+}
+
+output "key_name_generated" {
```
Comment (osterman) [inline output.tf]: We do not need this output. We want a single output as that will make integration easier.

File: main.tf [hunk 14]
```diff
@@ -12,7 +8,32 @@ module "label" {
   tags       = "${var.tags}"
 }
 
-resource "aws_key_pair" "default" {
+resource "aws_key_pair" "non_empty" {
```
Comment (osterman) [inline main.tf]: rename `non_empty` to `imported`

File: main.tf [hunk 15]
```diff
@@ -12,7 +8,32 @@ module "label" {
...
-  public_key = "${data.template_file.default.rendered}"
+  public_key = "${file("${var.ssh_public_key_path}/${module.label.id}.pub")}"
+}
+
+resource "tls_private_key" "default" {
+  count     = "${var.generate_ssh_key  == true ? 1 : 0}"
+  algorithm = "${var.ssh_key_algorithm}"
+}
+
+resource "aws_key_pair" "empty" {
```
Comment (osterman) [inline main.tf]: Rename `empty` to `generated`

File: output.tf (hunk 12 above)
Comment (SweetOps) [reply output.tf]: empty output with:
```terraform
output "key_name" {
  value = "${element(compact(aws_key_pair.imported.*.key_name, aws_key_pair.generated.*.key_name), 0)}"
}
```
non-empty output with:
```terraform
output "key_name" {
  value = "${element(compact(concat(aws_key_pair.imported.*.key_name, aws_key_pair.generated.*.key_name)), 0)}"
}
```

**Comment by comeanother** [comment]
Tested.
screenshot from 2017-09-06 20-14-42

**Pull request merged**